nucliadb_utils/src/nucliadb_utils/aiopynecone/client.py

```diff
@@ -47,37 +90,320 @@ async def create_index(self, name: str, dimension: int) -> str:
+        - `pagination_token`: to fetch the next page. The token is provided in the response
+           if there are more pages to fetch.
+        """
+        headers = {"Api-Key": self.api_key}
+        params = {"limit": str(limit)}
+        if prefix is not None:
+            params["prefix"] = prefix
+        if pagination_token is not None:
+            params["paginationToken"] = pagination_token
+        response = await self.http_session.get(
+            "/vectors/list",
+            headers=headers,
+            params=params,
+        )
+        raise_for_status(response)
+        return ListResponse.model_validate(response.json())
+
+    async def list_all(
+        self, prefix: Optional[str] = None, page_size: int = 100
+    ) -> AsyncGenerator[str, None]:
+        """
+        Iterate over all vector ids from the index in a paginated manner.
+        Params:
+        - `prefix`: to filter vectors by their id prefix.
+        - `page_size`: to control the number of vectors fetched in each page.
+        """
+        pagination_token = None
+        while True:
+            response = await self.list_page(
+                prefix=prefix, limit=page_size, pagination_token=pagination_token
+            )
+            for vector_id in response.vectors:
+                yield vector_id.id
+            if response.pagination is None:
+                break
+            pagination_token = response.pagination.next
+
+    @pinecone_observer.wrap({"type": "delete_all"})
+    async def delete_all(self):
+        """
+        Delete all vectors in the index.
+        """
+        headers = {"Api-Key": self.api_key}
+        payload = {"deleteAll": True, "ids": [], "namespace": ""}
+        response = await self.http_session.post("/vectors/delete", headers=headers, json=payload)
+        try:
+            raise_for_status(response)
+        except PineconeAPIError as err:
+            if err.http_status_code == 404 and err.code == 5:
+                # Namespace not found. No vectors to delete.
+                return
+            raise
+
+    @pinecone_observer.wrap({"type": "delete_by_id_prefix"})
+    async def delete_by_id_prefix(
+        self, id_prefix: str, batch_size: int = MAX_DELETE_BATCH_SIZE, max_parallel_batches: int = 1
+    ) -> None:
+        """ """
+        if batch_size > MAX_DELETE_BATCH_SIZE:
+            logger.warning(f"Batch size {batch_size} is too large. Limiting to {MAX_DELETE_BATCH_SIZE}.")
+            batch_size = MAX_DELETE_BATCH_SIZE
+
+        semaphore = asyncio.Semaphore(max_parallel_batches)
+
+        async def _delete_batch(batch):
+            async with semaphore:
+                await self.delete(ids=batch)
+
+        tasks = []
+        async for batch in self._async_batchify(self.list_all(prefix=id_prefix), size=batch_size):
+            tasks.append(asyncio.create_task(_delete_batch(batch)))
+
+        await asyncio.gather(*tasks)
+
+    def _batchify(self, iterable, size: int):
```

Comment:
Curiosity: in Python 3.12 there's a function called `bached` in `itertools` that does this